### PR TITLE
feat: generate certificate as part of EC2 App pattern

### PR DESCRIPTION
**Pull request by jacobwinch**
## What does this change?
Previously, users had to create their ACM certificate outside of `cdk` and then pass it in via a parameter. Certificate generation is now handled by the EC2 App pattern*.

*If the DNS record required to validate the certificate can be created in a Route 53 hosted zone, users do not need to do anything. If the aforementioned DNS record needs to be created in NS1, this must be created manually before stack creation will complete. We should automate this step too but I think it's a separate, larger piece of work.

## Does this change require changes to existing projects or CDK CLI?
No, this pattern is still under development.

## How to test
Existing unit tests have been updated.

## How can we measure success?
More resources are version controlled; less work needs to happen outside of pattern creation.

## Have we considered potential risks?
I think this is reasonably low risk. In order to reduce risk of keeping certificates under version control (and/or deploying them automatically) all certificates are created with `"DeletionPolicy": "Retain"`.

### CODE REVIEW

**Comment by akash1810** [comment]
> If the DNS record required to validate the certificate can be created in a Route 53 hosted zone, users do not need to do anything. If the aforementioned DNS record needs to be created in NS1, this must be created manually before stack creation will complete. We should automate this step too but I think it's a separate, larger piece of work.

It would be good to print warnings about this, though this might be the role of RiffRaff or other deployment tooling.

NS1 work is automated for some of our domains with [dns-validation-lambda](https://github.com/guardian/dns-validation-lambda).

File: src/constructs/acm/certificate.ts [hunk 1]
```diff
@@ -8,7 +8,8 @@ import type { GuStack } from "../core";
 import { AppIdentity } from "../core/identity";
 import type { GuMigratingResource } from "../core/migrating";
 
-export type GuCertificateProps = Record<Stage, GuDnsValidatedCertificateProps> & GuMigratingResource & AppIdentity;
+export type GuCertificateProps = Record<Stage, GuDnsValidatedCertificateProps> & GuMigratingResource;
+export type GuCertificatePropsWithApp = GuCertificateProps & AppIdentity;
```
Comment (akash1810) [inline src/constructs/acm/certificate.ts:12]: Feels a little horrible to make this split just to make the pattern props API simpler. The only alternative I can think of is to `GuEc2AppProps` to [`Omit`](https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys) the app field in `GuCertificateProps`, but not much better.

Might be one to address later once the AppIdentity mixin is written?

**Review by akash1810 — APPROVED** [review]

File: src/constructs/acm/certificate.ts (hunk 1 above)
Comment (jacobwinch) [reply src/constructs/acm/certificate.ts:12]: > The only alternative I can think of is to `GuEc2AppProps` to [`Omit`](https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys) the app field in `GuCertificateProps`, but not much better.

Yep, I tried this alternative too and decided that the approach in this PR was 'least-bad' 🙈

Comment (jacobwinch) [reply src/constructs/acm/certificate.ts:12]: > Might be one to address later once the AppIdentity mixin is written?

I've already found myself needing to do this twice today ([see also](https://github.com/guardian/cdk/pull/446/files#diff-e4e7eb9118f3261284cbc3ed695b73b19661de3cbb7b438e102a8293a50719d1R19-R20)). So it's definitely going to be worth thinking about a general solution for this.

**Comment by jacobwinch** [comment]
> NS1 work is automated for some of our domains with [dns-validation-lambda](https://github.com/guardian/dns-validation-lambda).

Oh that's great, I didn't know about this! Perhaps it's worth adding any missing domains to the lambda then? That way we can tell users 'it will just work' without needing to expose any of this complexity to them?

**Comment by akash1810** [comment]
> > NS1 work is automated for some of our domains with [dns-validation-lambda](https://github.com/guardian/dns-validation-lambda).
> 
> Oh that's great, I didn't know about this! Perhaps it's worth adding any missing domains to the lambda then? That way we can tell users 'it will just work' without needing to expose any of this complexity to them?

I think it purposefully only operates on the gutools domains and not on *.theguardian.com. Is this right @sihil ?

**Comment by jacobwinch** [comment]
Update: I've just tested this out with our new hosted zone and I can confirm that this creates the certificate and the appropriate CNAME automatically 🎉

**Pull request merged**